metomi/rose/config_editor/upgrade_controller.py

```diff
@@ -0,0 +1,329 @@
+        self.treewindow.show()
+        self.treewindow.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.NEVER)
+        columns = metomi.rose.config_editor.DIALOG_COLUMNS_UPGRADE
+        for i, title in enumerate(columns):
+            column = Gtk.TreeViewColumn()
+            column.set_title(title)
+            if self.treemodel.get_column_type(i) == GObject.TYPE_BOOLEAN:
+                cell = Gtk.CellRendererToggle()
+                cell.connect("toggled", self._handle_toggle_upgrade, i)
+                cell.set_property("activatable", True)
+            elif i == 2:
+                self._combo_cell = Gtk.CellRendererCombo()
+                self._combo_cell.set_property("has-entry", False)
+                self._combo_cell.set_property("editable", True)
+                self._combo_cell.connect(
+                    "changed", self._handle_change_version, 2
+                )
+                cell = self._combo_cell
+            else:
+                cell = Gtk.CellRendererText()
+            if i == len(columns) - 1:
+                column.pack_start(cell, True, True, 0)
+            else:
+                column.pack_start(cell, False, True, 0)
+            column.set_cell_data_func(cell, self._set_cell_data, i)
+            self.treeview.append_column(column)
+        self.treeview.connect("cursor-changed", self._handle_change_cursor)
+        self.treewindow.add(self.treeview)
+        self.window.vbox.pack_start(
+            self.treewindow,
+            expand=True,
+            fill=True,
+            padding=metomi.rose.config_editor.SPACING_PAGE,
+        )
+        label = Gtk.Label(label=metomi.rose.config_editor.DIALOG_LABEL_UPGRADE)
+        label.show()
+        self.window.vbox.pack_start(
+            label, True, True, metomi.rose.config_editor.SPACING_PAGE
+        )
+        button_hbox = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL)
+        button_hbox.show()
+        all_versions_toggle_button = Gtk.CheckButton(
+            label=metomi.rose.config_editor.DIALOG_LABEL_UPGRADE_ALL,
+            use_underline=False,
+        )
+        all_versions_toggle_button.set_active(self.use_all_versions)
+        all_versions_toggle_button.connect(
+            "toggled", self._handle_toggle_all_versions
+        )
+        all_versions_toggle_button.show()
+        button_hbox.pack_start(
+            all_versions_toggle_button,
+            expand=False,
+            fill=False,
+            padding=metomi.rose.config_editor.SPACING_SUB_PAGE,
+        )
+        self.window.vbox.pack_end(
+            button_hbox, expand=False, fill=False, padding=0
+        )
+        self.ok_button = self.window.action_area.get_children()[0]
+        self.window.set_focus(all_versions_toggle_button)
+        self.window.set_focus(self.ok_button)
+        self._set_ok_to_upgrade()
+        max_size = metomi.rose.config_editor.SIZE_MACRO_DIALOG_MAX
+        my_size = self.window.size_request()
```

Comment:
I'm struggling to navigate the GTK docs, but I don't think `get_size_request` returns the same type as it did at GTK2. I'm also not clear, but suspect it's marked for deprecation.
```suggestion
        my_size = self.window.get_size()
```